src/libsyntax/ext/tt/macro_rules.rs

```diff
@@ -70,6 +70,29 @@ impl<'a> ParserAnyMacro<'a> {
             } else if !parser.sess.source_map().span_to_filename(parser.token.span).is_real() {
                 e.span_label(site_span, "in this macro invocation");
             }
+            match kind {
+                AstFragmentKind::Ty => {
```

Comment:
Can you refactor this `match` and the one in `expand.rs` into a function and then make this match into:
```rust
            match kind {
                AstFragmentKind::Pat if macro_ident.name == sym::vec => {
                    e.span_label(
                        site_span,
                        "use a slice pattern here instead",
                    );
                    e.help("for more information, see https://doc.rust-lang.org/edition-guide/\
                              rust-2018/slice-patterns.html");
                }
                _ => foobar(&mut e, kind), // <-- The new function.
            }
```